Add Resources and Style to Popup

cat0363 opened the pull request:
<!--
 Hello, and thank you for your interest in contributing to the .NET MAUI Toolkit! 

 Before you submit please check that this work relates to one of the following:
 - Bug fix
    If you haven't yet opened an Issue that reports the bug in detail, provides a reproduction sample, and has been verified + reproduced by a member of the .NET MAUI Toolkit core team, please do that before submitting a Pull Request.
 - Feature/Proposal
    If you haven't yet submitted a Proposal that has been Championed by a .NET MAUI core team member, please instead open a Discussion at https://github.com/communitytoolkit/maui/discussions/new where we can discuss the pros/cons of the feature and its implementation. 
 Any PR submitted that does not fit with the above options will be closed.
 -->

In this PR, we will add Resources and Style to the Popup so that Style can be applied to the Popup.

 ### Description of Change ###

 <!-- Describe your changes here. This only needs to be brief as the linked issues below will already cover the detailed changes. -->

Since Popup is a class that inherits from Element and not from VisualElement, it does not have Resources and Style.
Modifications were made with reference to the following source code of .NET MAUI.

[src\Controls\src\Core\MergedStyle.cs]
[src\Controls\src\Core\Shell\NavigableElement.cs]
[src\Controls\src\Core\VisualElement\VisualElement.cs]

See above for the main implementation.

Inherited IResourcesProvider to be able to specify Resources for Popup.

[src\CommunityToolkit.Maui\Views\Popup\Popup.shared.cs]

    public partial class Popup : Element, IPopup, IWindowController, IPropertyPropagationController, IResourcesProvider, IStyleSelectable

I also added the following implementation:

[src\CommunityToolkit.Maui\Views\Popup\Popup.shared.cs]

    ResourceDictionary? resources;
    bool IResourcesProvider.IsResourcesCreated => resources != null;

    public ResourceDictionary Resources
    {
        get
        {
            if (resources != null)
            {
                return resources;
            }

            resources = new ResourceDictionary();
            ((IResourceDictionary)resources).ValuesChanged += OnResourcesChanged;
            return resources;
        }
        set
        {
            if (resources == value)
            {
                return;
            }

            OnPropertyChanging();
            if (resources != null)
            {
                ((IResourceDictionary)resources).ValuesChanged -= OnResourcesChanged;
            }

            resources = value;
            OnResourcesChanged(value);
            if (resources != null)
            {
                ((IResourceDictionary)resources).ValuesChanged += OnResourcesChanged;
            }

            OnPropertyChanged();
        }
    }

    internal override void OnParentResourcesChanged(IEnumerable<KeyValuePair<string, object>> values)
    {
        if (values == null)
        {
            return;
        }

        if (!((IResourcesProvider)this).IsResourcesCreated || Resources.Count == 0)
        {
            base.OnParentResourcesChanged(values);
            return;
        }

        var innerKeys = new HashSet<string>(StringComparer.Ordinal);
        var changedResources = new List<KeyValuePair<string, object>>();
        foreach (KeyValuePair<string, object> c in Resources)
        {
            innerKeys.Add(c.Key);
        }

        foreach (KeyValuePair<string, object> value in values)
        {
            if (innerKeys.Add(value.Key))
            {
                changedResources.Add(value);
            }
            else if (value.Key.StartsWith(Style.StyleClassPrefix, StringComparison.Ordinal))
            {
                var innerStyle = Resources[value.Key] as List<Style>;
                var parentStyle = value.Value as List<Style>;
                if (innerStyle is not null)
                {
                    var mergedClassStyles = new List<Style>(innerStyle);
                    if (parentStyle is not null)
                    {
                        mergedClassStyles.AddRange(parentStyle);
                    }
                    changedResources.Add(new KeyValuePair<string, object>(value.Key, mergedClassStyles));
                }
            }
        }
        if (changedResources.Count != 0)
        {
            OnResourcesChanged(changedResources);
        }
    }

The above is ported with reference to the VisualElement class of .NET MAUI.
[src\Controls\src\Core\VisualElement\VisualElement.cs]

This makes it possible to specify Resources in the Popup.

Next, I added the following to make it possible to specify global styles.Next, I added the following to make it possible to specify global styles.

[src\CommunityToolkit.Maui\Views\Popup\Popup.shared.cs]

    readonly MergedStyle mergedStyle;

    public Popup()
    {
        platformConfigurationRegistry = new Lazy<PlatformConfigurationRegistry<Popup>>(() => new(this));

        mergedStyle = new MergedStyle(GetType().BaseType, this);
    }

GetType().BaseType is specified as the first argument of the constructor of MergedStyle, because the style search target within the implementation of the MergedStyle class is the following class. 

    static readonly IList<Type> s_stopAtTypes = new List<Type> { typeof(View), typeof(Compatibility.Layout<>), typeof(VisualElement), typeof(NavigableElement), typeof(Element) };

I referenced the implementation below.
[src\Controls\src\Core\MergedStyle.cs]
[src\Controls\src\Core\Shell\NavigableElement.cs]

Since Popup inherits from Element and Element is the first hit when searching for a class to apply the style, the type of Popup's Base class is specified.

This makes it possible to specify global styles in the Popup.

Next, Inherited IStyleSelectable to allow styles to be applied to the Popup.

[src\CommunityToolkit.Maui\Views\Popup\Popup.shared.cs]

    public partial class Popup : Element, IPopup, IWindowController, IPropertyPropagationController, IResourcesProvider, IStyleSelectable

I also added the following implementation:

[src\CommunityToolkit.Maui\Views\Popup\Popup.shared.cs]

    public static readonly BindableProperty StyleProperty = BindableProperty.Create(nameof(Style), typeof(Style), typeof(Popup), default(Style), propertyChanged: (bindable, oldValue, newValue) => ((Popup)bindable).mergedStyle.Style = (Style)newValue);

    public Style Style
    {
        get { return (Style)GetValue(StyleProperty); }
        set { SetValue(StyleProperty, value); }
    }

    [System.ComponentModel.TypeConverter(typeof(ListStringTypeConverter))]
    public IList<string> StyleClass
    {
        get { return mergedStyle.StyleClass; }
        set { mergedStyle.StyleClass = value; }
    }

I ported it with reference to the following implementation of .NET MAUI.
[src\Controls\src\Core\VisualElement\VisualElement.cs]

The above implementation allows you to specify implicit styles, explicit styles, dynamic styles, and style classes.

 ### Linked Issues ###
 <!-- Provide links to issues here (#35 will link to issue number 35). Ensure that a GitHub issue was created for your bug/proposal and it has been approved/Championed. -->

 - Fixes #1345

 ### PR Checklist ###
 <!--
 Please check all the things you did here and double-check that you got it all, or state why you didn't do something.

 If anything is unclear please do ask :)
 -->
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [ ] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [ ] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pull/XYZ <!-- Replace XYZ with your docs PR #. The checkbox will be automatically checked once the docs PR has been merged -->


 ### Additional information ###

 <!-- 
 Please use this to aid the reviewer, this could include stating which platform(s) have been tested
 -->
 
**Windows results are with PR #1350 applied.**
Below is the validation result when specifying the global style.

[App.xaml]

    <Application xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
        xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
        xmlns:toolkit="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
        xmlns:local="clr-namespace:MauiComm_VerifyPopupStyle"
        x:Class="MauiComm_VerifyPopupStyle.App">
        <Application.Resources>
            <ResourceDictionary>
                <ResourceDictionary.MergedDictionaries>
                    <ResourceDictionary Source="Resources/Styles/Colors.xaml" />
                    <ResourceDictionary Source="Resources/Styles/Styles.xaml" />
                </ResourceDictionary.MergedDictionaries>

                <Style TargetType="toolkit:Popup">
                    <Setter Property="Size" Value="100, 100" />
                    <Setter Property="Color" Value="Red" />
                    <Setter Property="HorizontalOptions" Value="Start" />
                    <Setter Property="VerticalOptions" Value="End" />
                    <Setter Property="CanBeDismissedByTappingOutsideOfPopup" Value="False" />
                </Style>

            </ResourceDictionary>
        </Application.Resources>
    </Application>

[Popup1.xaml]

    <toolkit:Popup xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
        xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
        xmlns:toolkit="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
        x:Class="MauiComm_VerifyPopupStyle.Popup1">

        <Grid>
            <Grid.GestureRecognizers>
                <TapGestureRecognizer Tapped="gd_Tapped" />
            </Grid.GestureRecognizers>
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="Start" HorizontalOptions="Start" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="Start" HorizontalOptions="End" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="End" HorizontalOptions="Start" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="End" HorizontalOptions="End" BackgroundColor="Blue" />
        </Grid>

    </toolkit:Popup>

<table>
  <tr>
    <td>[Android]</td>
    <td>[iOS]</td>
    <td>[Windows]</td>
  </tr>
  <tr>
    <td><video controls width="250" src="https://github.com/CommunityToolkit/Maui/assets/125236133/fa990a8d-b12e-4129-bb0d-2ef99dbd6ce9" /></td>
    <td><video controls width="250" src="https://github.com/CommunityToolkit/Maui/assets/125236133/c0d36209-5916-4f7c-81c9-ecc4bd7ceb63" /></td>
    <td><video controls width="250" src="https://github.com/CommunityToolkit/Maui/assets/125236133/6ebe194e-1a62-48ca-bd9c-c812cd0fe6b2" /></td>
  </tr>
</table>

Below is the validation result when specifying an explicit style.

    <Application xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
        xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
        xmlns:toolkit="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
        xmlns:local="clr-namespace:MauiComm_VerifyPopupStyle"
        x:Class="MauiComm_VerifyPopupStyle.App">
        <Application.Resources>
            <ResourceDictionary>
                <ResourceDictionary.MergedDictionaries>
                    <ResourceDictionary Source="Resources/Styles/Colors.xaml" />
                    <ResourceDictionary Source="Resources/Styles/Styles.xaml" />
                </ResourceDictionary.MergedDictionaries>

                <Style x:Key="stlExplicitStyle" TargetType="toolkit:Popup">
                    <Setter Property="Size" Value="200, 100" />
                    <Setter Property="Color" Value="Yellow" />
                    <Setter Property="HorizontalOptions" Value="End" />
                    <Setter Property="VerticalOptions" Value="Start" />
                    <Setter Property="CanBeDismissedByTappingOutsideOfPopup" Value="False" />
                </Style>

            </ResourceDictionary>
        </Application.Resources>
    </Application>

[Popup2.xaml]

    <toolkit:Popup xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
        xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
        xmlns:toolkit="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
        x:Class="MauiComm_VerifyPopupStyle.Popup2" Style="{StaticResource stlExplicitStyle}">

        <Grid>
            <Grid.GestureRecognizers>
                <TapGestureRecognizer Tapped="gd_Tapped" />
            </Grid.GestureRecognizers>
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="Start" HorizontalOptions="Start" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="Start" HorizontalOptions="End" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="End" HorizontalOptions="Start" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="End" HorizontalOptions="End" BackgroundColor="Blue" />
        </Grid>

    </toolkit:Popup>

<table>
  <tr>
    <td>[Android]</td>
    <td>[iOS]</td>
    <td>[Windows]</td>
  </tr>
  <tr>
    <td><video controls width="250" src="https://github.com/CommunityToolkit/Maui/assets/125236133/96a9432e-0119-4ef5-98a0-44d27bf5f225" /></td>
    <td><video controls width="250" src="https://github.com/CommunityToolkit/Maui/assets/125236133/d1411e39-77d5-4a26-a4c8-99dd045bee88" /></td>
    <td><video controls width="250" src="https://github.com/CommunityToolkit/Maui/assets/125236133/0e05dafa-fb29-4e5b-a72e-879fef0f0184" /></td>
  </tr>
</table>

Below is the validation result when specifying the implicit style.

[Popup3.xaml]

    <toolkit:Popup xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
        xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
        xmlns:toolkit="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
        x:Class="MauiComm_VerifyPopupStyle.Popup3">
        <toolkit:Popup.Resources>
            <Style TargetType="toolkit:Popup">
                <Setter Property="Size" Value="100, 200" />
                <Setter Property="Color" Value="Green" />
                <Setter Property="HorizontalOptions" Value="Center" />
                <Setter Property="VerticalOptions" Value="Center" />
                <Setter Property="CanBeDismissedByTappingOutsideOfPopup" Value="False" />
            </Style>
        </toolkit:Popup.Resources>

        <Grid>
            <Grid.GestureRecognizers>
                <TapGestureRecognizer Tapped="gd_Tapped" />
            </Grid.GestureRecognizers>
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="Start" HorizontalOptions="Start" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="Start" HorizontalOptions="End" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="End" HorizontalOptions="Start" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="End" HorizontalOptions="End" BackgroundColor="Blue" />
        </Grid>

    </toolkit:Popup>

<table>
  <tr>
    <td>[Android]</td>
    <td>[iOS]</td>
    <td>[Windows]</td>
  </tr>
  <tr>
    <td><video controls width="250" src="https://github.com/CommunityToolkit/Maui/assets/125236133/251a4366-2caf-469d-a1e6-b2aa54843522" /></td>
    <td><video controls width="250" src="https://github.com/CommunityToolkit/Maui/assets/125236133/3b41eded-0e94-4eab-a98c-3eb2482bb1a6" /></td>
    <td><video controls width="250" src="https://github.com/CommunityToolkit/Maui/assets/125236133/b73e38fa-4d9b-4dee-816c-bacac75d6b26" /></td>
  </tr>
</table>

Below is the validation result when specifying the dynamic style.

    <toolkit:Popup xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
        xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
        xmlns:toolkit="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
        x:Class="MauiComm_VerifyPopupStyle.Popup4" Style="{DynamicResource stlDynamicStyle}">
        <toolkit:Popup.Resources>
            <Style x:Key="stlDynamicStyle" TargetType="toolkit:Popup">
                <Setter Property="Size" Value="100, 100" />
                <Setter Property="Color" Value="Pink" />
                <Setter Property="HorizontalOptions" Value="Center" />
                <Setter Property="VerticalOptions" Value="Center" />
                <Setter Property="CanBeDismissedByTappingOutsideOfPopup" Value="False" />
            </Style>
        </toolkit:Popup.Resources>

        <Grid>
            <Grid.GestureRecognizers>
                <TapGestureRecognizer Tapped="gd_Tapped" />
            </Grid.GestureRecognizers>
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="Start" HorizontalOptions="Start" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="Start" HorizontalOptions="End" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="End" HorizontalOptions="Start" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="End" HorizontalOptions="End" BackgroundColor="Blue" />
        </Grid>

    </toolkit:Popup>

<table>
  <tr>
    <td>[Android]</td>
    <td>[iOS]</td>
    <td>[Windows]</td>
  </tr>
  <tr>
    <td><video controls width="250" src="https://github.com/CommunityToolkit/Maui/assets/125236133/e8ab97ed-f94d-42af-b79e-bb3c5f36496f" /></td>
    <td><video controls width="250" src="https://github.com/CommunityToolkit/Maui/assets/125236133/8922857f-686c-442d-bd89-e1fd7ca1f30d" /></td>
    <td><video controls width="250" src="https://github.com/CommunityToolkit/Maui/assets/125236133/5916fa1c-d1db-4d0f-a304-6d05b6fc4a03" /></td>
  </tr>
</table>

Below is the validation result when specifying the style class.

    <toolkit:Popup xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
        xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
        xmlns:toolkit="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
        x:Class="MauiComm_VerifyPopupStyle.Popup5" StyleClass="PopupStyle">
        <toolkit:Popup.Resources>
            <Style TargetType="toolkit:Popup" Class="PopupStyle">
                <Setter Property="Size" Value="150, 150" />
                <Setter Property="Color" Value="Orange" />
                <Setter Property="HorizontalOptions" Value="Center" />
                <Setter Property="VerticalOptions" Value="End" />
                <Setter Property="CanBeDismissedByTappingOutsideOfPopup" Value="False" />
            </Style>
        </toolkit:Popup.Resources>

        <Grid>
            <Grid.GestureRecognizers>
                <TapGestureRecognizer Tapped="gd_Tapped" />
            </Grid.GestureRecognizers>
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="Start" HorizontalOptions="Start" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="Start" HorizontalOptions="End" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="End" HorizontalOptions="Start" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="End" HorizontalOptions="End" BackgroundColor="Blue" />
        </Grid>

    </toolkit:Popup>

<table>
  <tr>
    <td>[Android]</td>
    <td>[iOS]</td>
    <td>[Windows]</td>
  </tr>
  <tr>
    <td><video controls width="250" src="https://github.com/CommunityToolkit/Maui/assets/125236133/9ac01839-5f76-4b07-96ee-d5770b49b779" /></td>
    <td><video controls width="250" src="https://github.com/CommunityToolkit/Maui/assets/125236133/8cd05bab-0aa6-4413-b6f6-18f115fdf2f8" /></td>
    <td><video controls width="250" src="https://github.com/CommunityToolkit/Maui/assets/125236133/f15a5ec0-c77d-4b57-a2af-0d76cea01e34" /></td>
  </tr>
</table>

From the above, you can see that the style can be applied to the Popup as intended.
I would like to ask you to correct the document.

brminnick reviewed (changes requested):
FYI - a unit test is failing https://dev.azure.com/dotnet/CommunityToolkit/_build/results?buildId=97272&view=logs&j=3f96dcf5-6e1e-5485-3200-c557d5216be3&t=885110bc-af55-5c2d-c8fd-8677fc4411e5&l=49

```
 Failed CommunityToolkit.Maui.UnitTests.Views.PopupTests.NullColorThrowsArgumentNullException [8 ms]
  Error Message:
   System.NullReferenceException : Object reference not set to an instance of an object.
  Stack Trace:
     at Microsoft.Maui.Controls.MergedStyle.RegisterImplicitStyles()
   at Microsoft.Maui.Controls.MergedStyle..ctor(Type targetType, BindableObject target)
   at CommunityToolkit.Maui.Views.Popup..ctor() in /_/src/CommunityToolkit.Maui/Views/Popup/Popup.shared.cs:line 68
   at CommunityToolkit.Maui.UnitTests.Views.PopupTests.NullColorThrowsArgumentNullException() in /Users/runner/work/1/s/src/CommunityToolkit.Maui.UnitTests/Views/Popup/PopupTests.cs:line 190
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

cat0363 commented:
@brminnick , Thanks for the unit test results.
I'll try to investigate why the unit test failed.

cat0363 commented:
I modified the TargetType passed to the constructor of the MergedStyle class as follows.

[When the result obtained by GetType() is Popup]

    GetType()

[When the result obtained by GetType() is not Popup]

    GetType().BaseType

[src\CommunityToolkit.Maui\Views\Popup\Popup.shared.cs]

    public Popup()
    {
        platformConfigurationRegistry = new Lazy<PlatformConfigurationRegistry<Popup>>(() => new(this));

        var type = GetType();		
        mergedStyle = new MergedStyle(type == typeof(Popup) ? type : type.BaseType, this);
    }

When using the Popup class without inheriting it, when searching for implicit styles with the RegisterImplicitStyles 
method of the MergedStyle class, the Popup's BaseType was Element and the intended style could not be found, 
causing a NullReferenceException. If you don't inherit the Popup class and use it as it is, I modified it so that the 
TargetType is not the Popup's BaseType, but the Popup itself. This allows implicit styles to be explored as intended.

pictos commented on src/CommunityToolkit.Maui/Views/Popup/Popup.shared.cs:
```diff
@@ -40,13 +42,21 @@ public partial class Popup : Element, IPopup, IWindowController, IPropertyPropag
 	TaskCompletionSource<object?> resultTaskCompletionSource = new();
 	Window? window;
+	ResourceDictionary? resources;
+	bool IResourcesProvider.IsResourcesCreated => resources != null;
```
```suggestion
	bool IResourcesProvider.IsResourcesCreated => resources is not null;
```

pictos commented on src/CommunityToolkit.Maui/Views/Popup/Popup.shared.cs:
```diff
@@ -55,7 +65,8 @@ public Popup()
 	{
 		platformConfigurationRegistry = new Lazy<PlatformConfigurationRegistry<Popup>>(() => new(this));
 
-		VerticalOptions = HorizontalOptions = LayoutAlignment.Center;
```
Any particular reason to remove this line?

cat0363 replied:
@pictos , Thank you for your review.

I removed that code because LayoutAlignment.Center is set as the default value for VerticalOptionsProperty, HorizontalOptionsProperty. Also, if you explicitly specify VerticalOptions and HorizontalOptions before 
applying the style, the style won't be reflected, so I removed that code.

pictos replied:
I thought it would be overridden if the value were added after the style... Interesting, and how the popup is shown now? Those values were there to make sure the popup renderers at the center of the screen

cat0363 replied:
@pictos , I will answer your question.
If you have the code below, the behavior will be the same as when you explicitly specify the value of the Popup property in XAML, and the value specified explicitly will take precedence over the value specified in the style. 

    VerticalOptions = HorizontalOptions = LayoutAlignment.Center;

I verified it again, and if the above code exists, even if HorizontalOptions and VerticalOptions are specified on the style, the Popup will be displayed in the center of the screen regardless of their values.

Although it is unrelated to this case, as a result of verification, we will modify the targetType passed to the MergedStyle constructor as follows.

    mergedStyle = new MergedStyle(GetType(), this);

The reason is that when applying a global style to a class that inherits from Popup, the target style was not found.
The above corrections will be uploaded when the sample code is added.

Sorry for the late reply as I was dealing with another matter at work.

bijington replied:
When testing with styles were you making use of the `ApplyToDerivedTypes` property on `Style`?

cat0363 replied:
@bijington, Since the ApplyToDerivedTypesStyle property was not used during testing, I will specify it and verify it.
Thank you for your comment.

cat0363 replied:
I fixed the MergedStyle argument and uploaded it along with the sample page.

Among the things I tested, dynamic style inheritance was the only one that didn't work.
If you specify the BaseResourceKey property of Style, the style will not be applied.
Dynamic style inheritance is not available for Popups, but it is available for the Popup's content.

At least the following will work.

・Implicit Style
・Explicit Style
・Dynamic Style
・ApplyToDerivedTypes
・Style Inheritance
・Style Class

When creating a new Popup, it must be derived from the Popup class, so you can also 
specify the ApplyToDerivedTypes property of Style. 

>We don't have this luxury with Popups because a user can't just create a new instance of Popup, it has to be a derived type.
@bijington , Sorry, I don't understand the deeper meaning of the above.

cat0363 replied:
When debugging the dynamic style inheritance pattern, an exception occurred at the following location.

[src\CommunityToolkit.Maui\Views\Popup\Popup.shared.cs]

    public static readonly BindableProperty StyleProperty = BindableProperty.Create(nameof(Style), typeof(Style), typeof(Popup), default(Style), propertyChanged: (bindable, oldValue, newValue) => ((Popup)bindable).mergedStyle.Style = (Style)newValue);

The exception that occurred is as follows.

{System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Maui.Controls.Style.OnBasedOnResourceChanged(BindableObject bindable, Object oldValue, Object newValue) in D:\a\_work\1\s\src\Controls\src\Core\Style.cs:line 165
   at Microsoft.Maui.Controls.BindableObject.SetValueActual(BindableProperty property, BindablePropertyContext context, Object value, Boolean currentlyApplying, SetValueFlags attributes, Boolean silent) in D:\a\_work\1\s\src\Controls\src\Core\BindableObject.cs:line 535
   at Microsoft.Maui.Controls.BindableObject.SetValueCore(BindableProperty property, Object value, SetValueFlags attributes, SetValuePrivateFlags privateAttributes) in D:\a\_work\1\s\src\Controls\src\Core\BindableObject.cs:line 469
   at Microsoft.Maui.Controls.BindableObject.SetValueCore(BindableProperty property, Object value, SetValueFlags attributes) in D:\a\_work\1\s\src\Controls\src\Core\BindableObject.cs:line 401
   at Microsoft.Maui.Controls.Element.OnResourceChanged(BindableProperty property, Object value) in D:\a\_work\1\s\src\Controls\src\Core\Element.cs:line 634
   at Microsoft.Maui.Controls.Element.OnSetDynamicResource(BindableProperty property, String key) in D:\a\_work\1\s\src\Controls\src\Core\Element.cs:line 507
   at Microsoft.Maui.Controls.BindableObject.SetDynamicResource(BindableProperty property, String key, Boolean fromStyle) in D:\a\_work\1\s\src\Controls\src\Core\BindableObject.cs:line 361
   at Microsoft.Maui.Controls.BindableObject.SetDynamicResource(BindableProperty property, String key) in D:\a\_work\1\s\src\Controls\src\Core\BindableObject.cs:line 342
   at Microsoft.Maui.Controls.Style.Microsoft.Maui.Controls.IStyle.Apply(BindableObject bindable) in D:\a\_work\1\s\src\Controls\src\Core\Style.cs:line 104
   at CommunityToolkit.Maui.Views.MergedStyle2.SetStyle(IStyle implicitStyle, IList`1 classStyles, IStyle style) in C:\Users\tester\Documents\GitHub8\Maui\src\CommunityToolkit.Maui\Views\Popup\Popup.shared.cs:line 704
   at CommunityToolkit.Maui.Views.MergedStyle2.set_Style(IStyle value) in C:\Users\tester\Documents\GitHub8\Maui\src\CommunityToolkit.Maui\Views\Popup\Popup.shared.cs:line 496
   at CommunityToolkit.Maui.Views.Popup.<>c.<.cctor>b__87_0(BindableObject bindable, Object oldValue, Object newValue) in C:\Users\tester\Documents\GitHub8\Maui\src\CommunityToolkit.Maui\Views\Popup\Popup.shared.cs:line 53}

The source is .NET MAUI version 7.0.92.

If you catch the exception and apply the Style, it will work as intended, but that is an unwanted modification.

[src\CommunityToolkit.Maui\Views\Popup\Popup.shared.cs]

    public static readonly BindableProperty StyleProperty = BindableProperty.Create(nameof(Style), typeof(Style), typeof(Popup), default(Style), propertyChanged: (bindable, oldValue, newValue) => {
        try
        {
            ((Popup)bindable).mergedStyle.Style = (Style)newValue;
        }
        catch (NullReferenceException)
        {			
            ((Popup)bindable).mergedStyle.Style.Apply(bindable);
        }
    });

I tested it using the following style.

[App.xaml]

    <Application xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
        xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
        xmlns:popups="clr-namespace:CommunityToolkit.Maui.Sample.Views.Popups"
        xmlns:mct="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
        x:Class="CommunityToolkit.Maui.Sample.App">
        <Application.Resources>
            <ResourceDictionary>
                <ResourceDictionary.MergedDictionaries>
                    <ResourceDictionary Source="Resources/Styles/Colors.xaml" />
                    <ResourceDictionary Source="Resources/Styles/Styles.xaml" />
                </ResourceDictionary.MergedDictionaries>

                <Style x:Key="stlDynamicBaseStyle" TargetType="{x:Type popups:DynamicStyleInheritancePopup}">
                    <Setter Property="Size" Value="150,150" />
                </Style>
                <Style x:Key="stlDynamicInheritanceStyle1" TargetType="{x:Type popups:DynamicStyleInheritancePopup}" BasedOn="{StaticResource stlDynamicBaseStyle}">
                    <Setter Property="Color" Value="Purple" />
                    <Setter Property="VerticalOptions" Value="End" />
                </Style>
                <Style x:Key="stlDynamicInheritanceStyle2" TargetType="{x:Type popups:DynamicStyleInheritancePopup}" BaseResourceKey="stlDynamicInheritanceStyle1">
                    <Setter Property="Color" Value="Orange" />
                    <Setter Property="HorizontalOptions" Value="End" />
                    <Setter Property="CanBeDismissedByTappingOutsideOfPopup" Value="True" />
                </Style>
            </ResourceDictionary>
    </Application.Resources>
    </Application>

[DynamicStyleInheritancePopup.xaml]

    <mct:Popup xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
        xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
        xmlns:mct="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
        xmlns:popups="clr-namespace:CommunityToolkit.Maui.Sample.Views.Popups"
        x:Class="CommunityToolkit.Maui.Sample.Views.Popups.DynamicStyleInheritancePopup" Style="{StaticResource stlDynamicInheritanceStyle2}">
        <Grid VerticalOptions="Fill" HorizontalOptions="Fill" BackgroundColor="Transparent">
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="Start" HorizontalOptions="Start" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="Start" HorizontalOptions="End" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="End" HorizontalOptions="Start" BackgroundColor="Blue" />
            <Grid WidthRequest="10" HeightRequest="10" VerticalOptions="End" HorizontalOptions="End" BackgroundColor="Blue" />
        </Grid>
    </mct:Popup>

The execution results are as follows.

![Screenshot_1694504839](https://github.com/CommunityToolkit/Maui/assets/125236133/c0b0013b-b7a6-4e60-b842-31c726497075)

Although this is an expected result, it may be best to leave dynamic style inheritance non-guaranteed.

.NET MAUI 7.0.92 is already old, so it is unclear whether the issue will be resolved if verified with the latest source code.

bijington replied:
> I fixed the MergedStyle argument and uploaded it along with the sample page.
> 
> 
> 
> Among the things I tested, dynamic style inheritance was the only one that didn't work.
> 
> If you specify the BaseResourceKey property of Style, the style will not be applied.
> 
> Dynamic style inheritance is not available for Popups, but it is available for the Popup's content.
> 
> 
> 
> At least the following will work.
> 
> 
> 
> ・Implicit Style
> 
> ・Explicit Style
> 
> ・Dynamic Style
> 
> ・ApplyToDerivedTypes
> 
> ・Style Inheritance
> 
> ・Style Class
> 
> 
> 
> When creating a new Popup, it must be derived from the Popup class, so you can also 
> 
> specify the ApplyToDerivedTypes property of Style. 
> 
> 
> 
> >We don't have this luxury with Popups because a user can't just create a new instance of Popup, it has to be a derived type.
> 
> 
> 
> @bijington , Sorry, I don't understand the deeper meaning of the above.

I was trying compare how styles are used for existing controls like `Label`, etc. where a developer can create an instance of Label by adding it into their UI. They also have the ability to subclass Label let's call it `TitleLabel` and define a style for that. This is where `ApplyToDerivedTypes` makes sense. With `Popup` however a developer cannot simply add one to their UI and they have to create a subclass. Sorry I realise I sent you down the route of using ApplyToDerivedTypes but I am just now questioning whether it makes sense to make use of it. Part of me says we should and another part questions what should happen if that property is set to false, as this would result in a style definition that would never be used. Does that make sense?

cat0363 replied:
@bijington , Thank you for your reply. I understand what your comment means.

[Case 1]
If the value of the ApplyToDerivedTypes property is True, the style defined with the value of the ApplyToDerivedTypes property of True is applied unless a style that targets the derived type exists.

[Case 2]
If the value of the ApplyToDerivedTypes property is True and a style that targets a derived type is defined, the style that targets the derived type takes precedence.

[Case 3]
If the ApplyToDerivedTypes property is False, no styles are applied to derived types, so the value you explicitly set for the property has meaning.

Case 1 is used when you want to specify the same position, size, color, and closing behavior for all derived types. I think this is the case when you want to specify it in cases where it is a hassle to set Popup properties for each derived type one by one.

Case 2 is used when you want to explicitly apply a style when the same style has been applied to all derived types. For example, you might want to change the position of a Popup even though the size and color are the same.

Case 3 makes sense if you generate the Popup content in the code-behind without deriving the type from Popup. There are two ways to design the content of a Popup, but I think it is a case where the Popup is generated in the code-behind without inheriting the Popup class.

I do not choose Case 3 because I choose to design Popup content using XAML. I think Case 3 is necessary depending on how the screen is designed. This is just my personal opinion, but I don't think there is any problem with having the ApplyToDerivedTypes property.

cat0363 replied:
https://github.com/CommunityToolkit/Maui/pull/1351#discussion_r1322602255
@bijington , What do you think about this?
If dynamic style inheritance is used, no styles will be applied and an exception will be raised. 
I think something needs to be done, but what should I do?

cat0363 replied:
@bijington , It seems that the problem with Dynamic Style can be solved without any tricks.
That is to inherit IStyleElement to the Popup class. 

[src\Controls\src\Core\Style.cs]

    static void OnBasedOnResourceChanged(BindableObject bindable, object oldValue, object newValue)
    {
        Style style = (bindable as IStyleElement).Style;     // <= here
        if (style == null)
            return;
        if (!style._targets.TryGetValue(bindable, out var objectspecificity))
            return;

        style.UnApplyCore(bindable, (Style)oldValue);
        style.ApplyCore(bindable, (Style)newValue, (SetterSpecificity)objectspecificity);
    }

When I tried to cast the bindable in the above source code to IStyleElement, it became null.
This is because Popup does not inherit IStyleElement.
I will add Dynamic Style to the sample and upload the modified code by the end of today.
Please wait a moment now.

cat0363 replied:
@bijington , I just posted the code to add support for Dynami style inheritance.
I made the Popup class inherit IStyleElement. I also added a Dynamic style inheritance pattern to the sample.

bijington replied:
Looking good!

pictos commented on src/CommunityToolkit.Maui/Views/Popup/Popup.shared.cs:
```diff
@@ -179,6 +199,56 @@ public bool CanBeDismissedByTappingOutsideOfPopup
+	{
+		get
+		{
+			if (resources != null)
```
```suggestion
			if (resources is not null)
```

pictos commented on src/CommunityToolkit.Maui/Views/Popup/Popup.shared.cs:
```diff
@@ -179,6 +199,56 @@ public bool CanBeDismissedByTappingOutsideOfPopup
+			}
+
+			OnPropertyChanging();
+			if (resources != null)
```
```suggestion
			if (resources is not null)
```

pictos commented on src/CommunityToolkit.Maui/Views/Popup/Popup.shared.cs:
```diff
@@ -179,6 +199,56 @@ public bool CanBeDismissedByTappingOutsideOfPopup
+
+			resources = value;
+			OnResourcesChanged(value);
+			if (resources != null)
```
```suggestion
			if (resources is not null)
```

pictos commented on src/CommunityToolkit.Maui/Views/Popup/Popup.shared.cs:
```diff
@@ -301,4 +371,51 @@ static void OnColorChanged(BindableObject bindable, object oldValue, object newV
+
+	internal override void OnParentResourcesChanged(IEnumerable<KeyValuePair<string, object>> values)
+	{
+		if (values == null)
```
```suggestion
		if (values is null)
```

pictos reviewed (commented):
Thanks for this PR, looks good in general, I believe this one will be an easy one to merge. Can you add a sample page showing of this feature?

cat0363 commented:
> Thanks for this PR, looks good in general, I believe this one will be an easy one to merge. Can you add a sample page showing of this feature?

I would like to think about a sample using the creation of the sample project as a reference.
Defining implicit styles will affect other Popups in the sample, so I would like to create a sample that applies 
styles to classes that inherit from Popup.

bijington commented:
> > Thanks for this PR, looks good in general, I believe this one will be an easy one to merge. Can you add a sample page showing of this feature?
> 
> 
> 
> I would like to think about a sample using the creation of the sample project as a reference.
> 
> Defining implicit styles will affect other Popups in the sample, so I would like to create a sample that applies 
> 
> styles to classes that inherit from Popup.

It would be great if would could show off some implicit and explicit styles. I'd be happy to assist with some of this work if you need?

cat0363 commented:
> > > Thanks for this PR, looks good in general, I believe this one will be an easy one to merge. Can you add a sample page showing of this feature?
> > 
> > 
> > I would like to think about a sample using the creation of the sample project as a reference.
> > Defining implicit styles will affect other Popups in the sample, so I would like to create a sample that applies
> > styles to classes that inherit from Popup.
> 
> It would be great if would could show off some implicit and explicit styles. I'd be happy to assist with some of this work if you need?

@bijington , I plan to add the following to the Popup Style sample page.

・Implicit Style
・Explicit Style
・Dynamic Style
・ApplyToDerivedTypes
・Style Inheritance
・Dynamic Style Inheritance
・Style Class

I'll upload it tomorrow, along with the code fixes, as soon as it's completed.

Tomorrow, if there are any missing patterns after uploading, I would like to ask for your help.

pictos commented on samples/CommunityToolkit.Maui.Sample/CommunityToolkit.Maui.Sample.csproj:
```diff
@@ -76,6 +76,51 @@
+      <DependentUpon>StyleInheritancePopup.xaml</DependentUpon>
+    </Compile>
+    <Compile Update="Views\Popups\ExplicitStylePopup.xaml.cs">
+      <DependentUpon>ExplicitStylePopup.xaml</DependentUpon>
+    </Compile>
+    <Compile Update="Views\Popups\ImplicitStylePopup.xaml.cs">
+      <DependentUpon>ImplicitStylePopup.xaml</DependentUpon>
+    </Compile>
+    <Compile Update="Views\Popups\DynamicStylePopup.xaml.cs">
+      <DependentUpon>DynamicStylePopup.xaml</DependentUpon>
+    </Compile>
+    <Compile Update="Views\Popups\StyleClassPopup.xaml.cs">
+      <DependentUpon>StyleClassPopup.xaml</DependentUpon>
+    </Compile>
+  </ItemGroup>
+
+  <ItemGroup>
+    <MauiXaml Update="Pages\Views\Popup\PopupStylePage.xaml">
+      <Generator>MSBuild:Compile</Generator>
+    </MauiXaml>
+    <MauiXaml Update="Views\Popups\ApplyToDerivedTypesPopup.xaml">
+      <Generator>MSBuild:Compile</Generator>
+    </MauiXaml>
+    <MauiXaml Update="Views\Popups\StyleInheritancePopup.xaml">
+      <Generator>MSBuild:Compile</Generator>
+    </MauiXaml>
+    <MauiXaml Update="Views\Popups\ExplicitStylePopup.xaml">
+      <Generator>MSBuild:Compile</Generator>
+    </MauiXaml>
+    <MauiXaml Update="Views\Popups\ImplicitStylePopup.xaml">
+      <Generator>MSBuild:Compile</Generator>
+    </MauiXaml>
+    <MauiXaml Update="Views\Popups\DynamicStylePopup.xaml">
+      <Generator>MSBuild:Compile</Generator>
+    </MauiXaml>
+    <MauiXaml Update="Views\Popups\StyleClassPopup.xaml">
+      <Generator>MSBuild:Compile</Generator>
+    </MauiXaml>
+  </ItemGroup>
+
```
You can delete those lines

```suggestion
  
```

cat0363 replied:
> You can delete those lines
@pictos , Thank you for your advice.

brminnick reviewed (approved):
Thanks @cat0363! 

@pictos + @bijington  - I see an [open conversation here](https://github.com/CommunityToolkit/Maui/pull/1351/files#r1320881194). Do we still have open action items before merging this PR?

bijington commented on samples/CommunityToolkit.Maui.Sample/App.xaml:
```diff
@@ -1,12 +1,53 @@
                 <ResourceDictionary Source="Resources/Styles/Styles.xaml" />
             </ResourceDictionary.MergedDictionaries>
+
+            <Style x:Key="ImplicitPopupStyle" TargetType="{x:Type popups:ImplicitStylePopup}">
```
I don't know why I have it in my head the the implicit and explicit options are the other way round. You have to explicitly use this one via `Style="{StaticResource ImplicitPopupStyle}"` if I am correct. Whereas the current `ExplicitStylePopup` is implicitly applied without having to define the `Style` property. Is that correct?

cat0363 replied:
You are right. I was naming the styles backwards.
I'll fix it and upload it again.

bijington commented on samples/CommunityToolkit.Maui.Sample/CommunityToolkit.Maui.Sample.csproj:
```diff
@@ -76,6 +76,18 @@
     <ProjectReference Include="..\..\src\CommunityToolkit.Maui.Maps\CommunityToolkit.Maui.Maps.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
```
We don't need this section

cat0363 replied:
Thank you for pointing it out. Delete the above.

bijington commented on samples/CommunityToolkit.Maui.Sample/Pages/Views/Popup/StylePopupPage.xaml:
```diff
@@ -0,0 +1,56 @@
+                Text="Implicit Style"
+                BackgroundColor="Blue"
+                TextColor="White"
+                Command="{Binding DisplayPopup1Command}"/>
```
It might be nicer to name these commands so that they provide clearer context to what they are doing, e.g. `DisplayImplicitStylePopupCommand`, etc.

cat0363 replied:
I agree with your opinion. Change the name of the command to something easier to understand.

bijington commented on samples/CommunityToolkit.Maui.Sample/ViewModels/Views/ViewsGalleryViewModel.cs:
```diff
@@ -31,6 +31,7 @@ public ViewsGalleryViewModel()
 			SectionModel.Create<PopupLayoutAlignmentViewModel>("Popup Layout Page", Colors.Red, "Popup.Content demonstrated using different layouts"),
 			SectionModel.Create<ShowPopupInOnAppearingPageViewModel>("Show Popup in OnAppearing", Colors.Red, "Proves that we now support showing a popup before the platform is even ready."),
 			SectionModel.Create<SemanticOrderViewPageViewModel>("Semantic Order View", Colors.Red, "SemanticOrderView allows developers to indicate the focus order of visible controls when a user is navigating via TalkBack (Android), VoiceOver (iOS) or Narrator (Windows)."),
+			SectionModel.Create<StylePopupViewModel>("Popup Style Page", Colors.Red, "A page demonstrating Popup Style."),
```
```suggestion
			SectionModel.Create<StylePopupViewModel>("Popup Style Page", Colors.Red, "A page demonstrating how Popups can be styled in a .NET MAUI application."),
```

bijington commented on src/CommunityToolkit.Maui/Views/Popup/Popup.shared.cs:
```diff
@@ -40,22 +42,31 @@ public partial class Popup : Element, IPopup, IWindowController, IPropertyPropag
 		platformConfigurationRegistry = new Lazy<PlatformConfigurationRegistry<Popup>>(() => new(this));
-
-		VerticalOptions = HorizontalOptions = LayoutAlignment.Center;
+		((IResourceDictionary)resources).ValuesChanged += OnResourcesChanged;
```
I don't have the code open right now to inspect the .NET MAUI implementation (in case it uses WeakEventManager) but do we need to unsubscribe from this?

cat0363 replied:
I write the unsubscribe handling in the OnClosed method.

cat0363 replied:
@brminnick , Thank you for fixing this.
I made sure you fixed it.

bijington commented on samples/CommunityToolkit.Maui.Sample/Views/Popups/ApplyToDerivedTypesPopup.xaml:
```diff
@@ -0,0 +1,22 @@
+             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
+             xmlns:mct="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
+             x:Class="CommunityToolkit.Maui.Sample.Views.Popups.ApplyToDerivedTypesPopup">
+    <mct:Popup.Resources>
```
I wonder if there is some value in adding xml comments to inform the reader what each style based scenario is covering as we want to use this as learning material. It could be something like:

```xaml
<!-- This popup demonstrates how the ApplyToDerivedTypes property can allow for Popup implementations that inherit from Popup to still use a common Style definition. -->
```

cat0363 replied:
I think it would be easier to understand if there was an explanation in the comments for each style.
Is it possible for you to write comments for each style?

bijington replied:
Yep I would be happy to take this on 👍

bijington replied:
I've opted to add XML docs instead. As it will hopefully be more discoverable.

cat0363 replied:
@bijington , Thank you for your willingness to accept. 
Thank you for your help.

bijington reviewed (changes requested):
Sorry for the delay in responding to this @cat0363. Thank you for the work. I have made some suggestions but please let me know what you think?

bijington commented on src/CommunityToolkit.Maui/Views/Popup/Popup.shared.cs:
```diff
@@ -352,6 +352,10 @@ internal override void OnParentResourcesChanged(IEnumerable<KeyValuePair<string,
 		((IPopup)this).OnClosed(result);
+		if (resources is not null)
+		{
+			((IResourceDictionary)resources).ValuesChanged -= OnResourcesChanged;
```
Do we need to cast here? If so it might be clearer to modify the if statement to pattern match e.g.

```csharp
if (resources is IResourceDictionary resourceDictionary)
{
    resourceDictionary.ValuesChanged -= OnResourcesChanged;
}
```

What do you think?

bijington replied:
Oh I see the interface is internal and we are casting in the constructor so we are consistent. Maybe we ignore what I just said as this will catch issues if the API changes

bijington reviewed (changes requested):
@cat0363 this is looking great! Just a small point on the casting. Otherwise I think we are good!

Thank you so much for this!

bijington reviewed (dismissed):
I think we are good! Thanks so much @cat0363

brminnick commented:
Awesome!! Congrats @cat0363! 

The next step is to write the documentation for the new APIs! I've temporarily added the https://github.com/CommunityToolkit/Maui/labels/pending%20documentation and https://github.com/CommunityToolkit/Maui/labels/do%20not%20merge labels to ensure we don't accidentally merge this PR before the docs PR has been approved. 

You can open a PR on the MicrosoftDocs repo here:
https://github.com/MicrosoftDocs/CommunityToolkit/pulls

I'll defer to @bijington to help with any questions you may have 🙌

bijington commented:
@cat0363 how do you feel about doing the docs? If you like I am more than happy to take on the effort

cat0363 commented:
@bijington , I wish I could write a document, but I'm embarrassed to say that I don't have the literary talent to write it in fluent English that anyone can understand. If possible, I would like to ask @bijington , who has published an excellent book, to create this document. I read your wonderful book. 
Please let me know if there is anything I can do to help, such as providing screen captures or samples.

> @cat0363 how do you feel about doing the docs? If you like I am more than happy to take on the effort

@brminnick , Thanks for explaining what the labels mean.

bijington commented:
@cat0363 poease don't be embarrassed, I will happily do it. And if you ever need assistance in this area I'm always willing to help.

And wow! Thank you so much for firstly buying the book and secondly for the kind words. I can't tell you how much I appreciate it!

cat0363 commented:
@bijington , Thank you for taking on the task of creating this document.
Your books contributed to improving my skills.

brminnick reviewed (approved):
Thanks @cat0363 and @bijington!!

cat0363 commented:
@bijington , Thank you for creating the document.
Thank you everyone for the code reviews.
I'm glad that this PR was featured on Standup.
